Adding interactivity (#104)
reactjs/sr.react.dev: src/content/learn/adding-interactivity.md

---
title: Dodavanje interaktivnosti
---

<Intro>

> Neke stvari na ekranu se update-uju kao odgovor na korisnički unos. Na primer, klik na galeriju slika menja aktivnu sliku. U React-u, podaci koji se menjaju tokom vremena nazivaju se *state.* Možete dodati state bilo kojoj component-i i update-ovati je po potrebi. U ovom poglavlju naučićete kako da pišete component-e koji obrađuju interakcije, update-uju svoj state i prikazuju različite output-e tokom vremena.

Reviewer: ```suggestion
Neke stvari na ekranu se update-uju kao odgovor na korisnički input. Na primer, klik na galeriju slika menja aktivnu sliku. U React-u, podaci koji se menjaju tokom vremena nazivaju se *state.* Možete dodati state bilo kojoj component-i i update-ovati je po potrebi. U ovom poglavlju naučićete kako da pišete component-e koji obrađuju interakcije, update-uju svoj state i prikazuju različite output-e tokom vremena.
```
I would rather leave english term here, same as we did in other articles.

</Intro>

<YouWillLearn isChapter={true}>

> * [Kako da rukujette event-ima koje pokrene korisnik](/learn/responding-to-events)

Reviewer: ```suggestion
* [Kako da rukujete event-ima koje pokrene korisnik](/learn/responding-to-events)
```

* [Kako da učinite da component-e „pamte” informacije koristeći state](/learn/state-a-components-memory)
* [Kako React update-uje UI u dve faze](/learn/render-and-commit)
* [Zašto se state ne update-uje odmah nakon promene](/learn/state-as-a-snapshot)
* [Kako da složite seriju state update-ova](/learn/queueing-a-series-of-state-updates)
* [Kako da update-ujete objekat u state-u](/learn/updating-objects-in-state)
* [Kako da update-ujete array u state-u](/learn/updating-arrays-in-state)

</YouWillLearn>

## Odgovaranje na event-e {/*responding-to-events*/}

React vam omogućava da dodate *event handler*-e u vaš JSX. Event handler-i su vaše sopstvene funkcije koje će se pokrenuti kao odgovor na korisničke interakcije, poput klika, prelaženja mišem, fokusiranja na form input-e i tako dalje.

> Ugrađene component-e poput `<button>` podržavaju samo ugrađene browser event-e poput `onClick`. Međutim, možete kreirati i sopstvene component-e i njihovim event handler prop-ovima dati bilo koja aplikacijski specifična imena koja želite.

Reviewer: ```suggestion
Ugrađene component-e poput `<button>` podržavaju samo ugrađene browser event-e poput `onClick`. Međutim, možete kreirati i sopstvene component-e i njihovim event handler prop-ovima dati bilo koja aplikacijski specifična imena.
```
I think we don't need to specify `koja želite` at the end of this sentence.

<LearnMore path="/learn/responding-to-events">

Pročitajte **[Odgovaranje na Event-e](/learn/responding-to-events)** da biste naučili kako da dodate event handler-e.

</LearnMore>

## State: memorija component-e {/*state-a-components-memory*/}

> Component-e često moraju da promene ono što je prikazano na ekranu kao rezultat neke interakcije. Kucanje u formu treba da update-uje polje za unos, klik na "sledeće" u karuselu slika treba da promeni prikazanu sliku, klik na "kupi" stavlja proizvod u korpu. Component-e moraju da "pamte" određene stvari: trenutnu vrednost unosa, trenutnu sliku, korpu za kupovinu. U React-u, ova vrsta memorije specifična za component-e naziva se *state.*

Reviewer: ```suggestion
Component-e često moraju da promene ono što je prikazano na ekranu kao rezultat neke interakcije. Kucanje u formu treba da update-uje polje za unos, klik na "sledeće" u carousel-u slika treba da promeni prikazanu sliku, klik na "kupi" stavlja proizvod u korpu. Component-e moraju da "pamte" određene stvari: trenutnu vrednost unosa, trenutnu sliku, korpu za kupovinu. U React-u, ova vrsta memorije specifična za component-e naziva se *state.*
```
I would rather use english term than transliterating it to Serbian.

> Možete dodati state component-i pomoću Hook-a [`useState`](/reference/react/useState). *Hook-ovi* su specijalne funkcije koje omogućavaju vašim component-ama da koriste React funkcionalnosti (state je jedna od tih funkcionalnosti). Hook `useState` vam omogućava da deklarišete varijablu state-a. On prima inicijalni state i vraća par vrednosti: trenutni state i funkciju za setovanje state-a koja vam omogućava da ga update-ujete.

Reviewer: ```suggestion
Možete dodati state component-i pomoću Hook-a [`useState`](/reference/react/useState). *Hook-ovi* su specijalne funkcije koje omogućavaju vašim component-ama da koriste React funkcionalnosti (state je jedna od tih funkcionalnosti). Hook `useState` vam omogućava da deklarišete promenljivu state-a. On prima inicijalni state i vraća par vrednosti: trenutni state i funkciju za setovanje state-a koja vam omogućava da ga update-ujete.
```

<LearnMore path="/learn/state-a-components-memory">

Pročitajte **[State: Memorija Component-e](/learn/state-a-components-memory)** kako biste naučili kako da zapamtite vrednost i update-ujete je prilikom interakcije.

</LearnMore>

## Render i commit {/*render-and-commit*/}

> Pre nego što se vaše component-e prikažu na ekranu, React mora da ih render-uje. Razumevanje koraka u ovom procesu pomoći će vam da razmislite o tome kako se vaš kod izvršava i da objasnite njegovo ponašanje.

Reviewer: ```suggestion
Pre nego što se vaše component-e prikažu na ekranu, React mora da ih renderuje. Razumevanje koraka u ovom procesu pomoći će vam da razmislite o tome kako se vaš kod izvršava i da objasnite njegovo ponašanje.
```

Zamislite da su vaše component-e kuvari u kuhinji, koji pripremaju ukusna jela od sastojaka. U ovom scenariju, React je konobar koji prenosi narudžbine od gostiju i donosi im njihova jela. Ovaj proces naručivanja i posluživanja UI-a ima tri koraka:

> 1. **Pokretanje** render-a (prenos narudžbine gosta u kuhinju)

Reviewer: ```suggestion
1. **Pokretanje** renderovanja (prenos narudžbine gosta u kuhinju)
```

> 2. **Render-ovanje** component-e (priprema narudžbine u kuhinji)

Reviewer: ```suggestion
2. **Renderovanje** component-e (priprema narudžbine u kuhinji)
```

3. **Commit-ovanje** u DOM (postavljanje narudžbine na sto)

<IllustrationBlock sequential>
  <Illustration caption="Trigger" alt="React kao konobar u restoranu, prenosi narudžbine od korisnika i dostavlja ih u Component-u Kitchen." src="/images/docs/illustrations/i_render-and-commit1.png" />
  <Illustration caption="Render" alt="Šef kuhinje za Card daje React-u svežu Card component-u." src="/images/docs/illustrations/i_render-and-commit2.png" />

>   <Illustration caption="Commit" alt="React donosi Card korisniku za njihov sto." src="/images/docs/illustrations/i_render-and-commit3.png" />

Reviewer: ```suggestion
  <Illustration caption="Commit" alt="React donosi Card korisniku za njegov sto." src="/images/docs/illustrations/i_render-and-commit3.png" />
```

</IllustrationBlock>

<LearnMore path="/learn/render-and-commit">

Pročitajte **[Render i Commit](/learn/render-and-commit)** kako biste naučili lifecycle jednog UI update-a.

</LearnMore>

## State kao snapshot {/*state-as-a-snapshot*/}

> Za razliku od običnih JavaScript varijabli, React state se ponaša više kao snapshot. Njegovo postavljanje ne menja već postojeću state varijablu, već umesto toga pokreće re-render. Ovo može biti iznenađujuće na početku!

Reviewer: ```suggestion
Za razliku od običnih JavaScript promenljivih, React state se ponaša više kao snapshot. Njegovo postavljanje ne menja već postojeću state promenljivu, već umesto toga ponovo pokreće renderovanje. Ovo može biti iznenađujuće na početku!
```

```js
console.log(count);  // 0

> setCount(count + 1); // Request a re-render with 1

Reviewer: We should translate comments as well

</Sandpack>

> [State kao Snapshot](/learn/state-as-a-snapshot) objašnjava zašto se ovo dešava. Postavljanje state-a traži novo render-ovanje, ali ne menja njegovu vrednost u kodu koji se već izvršava. Tako `score` nastavlja da ima vrednost `0` odmah nakon što pozovete `setScore(score + 1)`.

Reviewer: ```suggestion
[State kao Snapshot](/learn/state-as-a-snapshot) objašnjava zašto se ovo dešava. Postavljanje state-a traži novo renderovanje, ali ne menja njegovu vrednost u kodu koji se već izvršava. Tako `score` nastavlja da ima vrednost `0` odmah nakon što pozovete `setScore(score + 1)`.
```

<LearnMore path="/learn/queueing-a-series-of-state-updates">

Pročitajte **Redosled serije state update-ova](/learn/queueing-a-series-of-state-updates)** kako biste naučili kako da postavite redosled update-ova state-a.

</LearnMore>

## Update-ovanje objekata u state-u {/*updating-objects-in-state*/}

State može sadržati bilo koju vrstu JavaScript vrednosti, uključujući objekte. Međutim, ne bi trebalo direktno menjati objekte i array-e koje držite u React state-u. Umesto toga, kada želite da update-ujete objekat ili array, potrebno je da kreirate novi (ili napravite kopiju postojećeg) i zatim update-ujete state kako bi koristio tu kopiju.

> Obično ćete koristiti `...` spread sintaksu za kopiranje objekata i array-a koje želite da promenite. Na primer, update-ovanje ugnježdenog objekta može izgledati ovako:

Reviewer: ```suggestion
Obično ćete koristiti `...` spread sintaksu za kopiranje objekata i array-a koje želite da ih promenite. Na primer, update-ovanje ugnježdenog objekta može izgledati ovako:
```

> <Sandpack>

Reviewer: We should translate coda examples as well
